tango/step_graph.py

```diff
@@ -1,107 +1,160 @@
+                        f"step {nd_step.name}. This will produce confusing results."
+                    )
+
+    @classmethod
+    def _find_step_dependencies(cls, o: Any) -> Set[str]:
+        dependencies: Set[str] = set()
+        if isinstance(o, (list, tuple, set)):
+            for item in o:
+                dependencies = dependencies | cls._find_step_dependencies(item)
+        elif isinstance(o, dict):
+            if set(o.keys()) == {"type", "ref"} and o["type"] == "ref":
+                dependencies.add(o["ref"])
+            else:
+                for value in o.values():
+                    dependencies = dependencies | cls._find_step_dependencies(value)
+        elif o is not None and not isinstance(o, (bool, str, int, float)):
+            raise ValueError(o)
+        return dependencies
 
-            raise ConfigurationError("Invalid step graph:\n- " + "\n- ".join(err_msgs))
+    @classmethod
+    def _replace_step_dependencies(cls, o: Any, existing_steps: Mapping[str, Step]) -> Any:
+        if isinstance(o, (list, tuple, set)):
+            return o.__class__(cls._replace_step_dependencies(i, existing_steps) for i in o)
+        elif isinstance(o, dict):
+            if set(o.keys()) == {"type", "ref"} and o["type"] == "ref":
+                return existing_steps[o["ref"]]
+            else:
+                return {
+                    key: cls._replace_step_dependencies(value, existing_steps)
+                    for key, value in o.items()
+                }
+        elif o is not None and not isinstance(o, (bool, str, int, float)):
+            raise ValueError(o)
+        return o
 
-    def __getitem__(self, key: Union[str, int]) -> StepStub:  # type: ignore[override]
+    def __getitem__(self, name: str) -> Step:
         """
-        Get the stub corresponding to ``key``.
+        Get the step with the given name.
         """
-        if isinstance(key, str):
-            return self._ordered_steps[key]
-        else:
-            return self._ordered_steps[list(self._ordered_steps.keys())[key]]
+        return self.parsed_steps[name]
 
     def __len__(self) -> int:
         """
         The number of steps in the experiment.
         """
-        return len(self._ordered_steps)
+        return len(self.parsed_steps)
 
-    @staticmethod
-    def _parse_direct_step_dependencies(o: Any) -> Set[str]:
-        dependencies: Set[str] = set()
-        if isinstance(o, (list, tuple, set)):
-            for item in o:
-                dependencies = dependencies | StepGraph._parse_direct_step_dependencies(item)
-        elif isinstance(o, dict):
-            if set(o.keys()) == {"type", "ref"}:
-                dependencies.add(o["ref"])
+    def __iter__(self) -> Iterator[str]:
+        """
+        The names of the steps in the experiment.
+        """
+        return iter(self.parsed_steps)
+
+    def serialized(self) -> List[Step]:
```

Comment:
This name is confusing to me because it could also be interpreted as "serialize" in the context of serialization/deserialization.

Reply:
What should we call it then? "`toposort()`"?

Reply:
I was thinking `ordered_steps()`

Reply:
That's not super precise, but good enough. 4e9910d4fe4cdff43e2d1e2a85ebc2c9328fe658